File: client-sdks/java/sdk/src/test/java/com/vertexcache/sdk/comm/GcmCryptoHelperTest.java
/*
 * Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.vertexcache.sdk.comm;

import com.vertexcache.sdk.comm.GcmCryptoHelper;
import org.junit.jupiter.api.Test;

import java.security.SecureRandom;
import java.util.Base64;

import static org.junit.jupiter.api.Assertions.*;

public class GcmCryptoHelperTest {

    private final byte[] key = new byte[32]; // 256-bit AES key (zeroed for test)
    private final byte[] message = "VertexCache secure payload".getBytes();

    @Test
    public void testEncryptDecryptRoundTrip() throws Exception {
        byte[] encrypted = GcmCryptoHelper.encrypt(message, key);
        assertNotNull(encrypted);
        assertTrue(encrypted.length > message.length);

        byte[] decrypted = GcmCryptoHelper.decrypt(encrypted, key);
        assertArrayEquals(message, decrypted);
    }

    @Test
    public void testDecryptFailsOnModifiedCiphertext() throws Exception {
        byte[] encrypted = GcmCryptoHelper.encrypt(message, key);
        encrypted[encrypted.length - 1] ^= 0x01; // flip last byte

        assertThrows(Exception.class, () -> GcmCryptoHelper.decrypt(encrypted, key));
    }

    @Test
    public void testDecryptFailsIfTooShort() {
        byte[] tooShort = new byte[5];
        assertThrows(IllegalArgumentException.class, () -> GcmCryptoHelper.decrypt(tooShort, key));
    }

    @Test
    public void testBase64KeyEncodeDecodeRoundTrip() {
        String base64 = GcmCryptoHelper.encodeBase64Key(key);
        byte[] decoded = GcmCryptoHelper.decodeBase64Key(base64);

        assertArrayEquals(key, decoded);
    }

    @Test
    public void testGenerateBase64Key() throws Exception {
        String base64Key = GcmCryptoHelper.generateBase64Key();
        assertNotNull(base64Key);

        byte[] decoded = GcmCryptoHelper.decodeBase64Key(base64Key);
        assertEquals(32, decoded.length); // 256-bit
    }

    @Test
    public void testEncryptGeneratesUniqueIVs() throws Exception {
        byte[] key = new byte[32];
        SecureRandom random = new SecureRandom();
        random.nextBytes(key);

        byte[] c1 = GcmCryptoHelper.encrypt(message, key);
        byte[] c2 = GcmCryptoHelper.encrypt(message, key);

        assertNotEquals(Base64.getEncoder().encodeToString(c1), Base64.getEncoder().encodeToString(c2));
    }
}

File: client-sdks/csharp/tests/VertexCacheSdk.Tests/Comm/GcmCryptoHelperTest.cs
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------------

using System;
using System.Security.Cryptography;
using System.Text;
using VertexCacheSdk.Comm;
using Xunit;

namespace VertexCacheSdk.Tests.Comm
{
    public class GcmCryptoHelperTest
    {
        private readonly byte[] key = new byte[32]; // 256-bit all-zero key
        private readonly byte[] message = Encoding.UTF8.GetBytes("VertexCache secure payload");

        [Fact]
        public void EncryptDecryptRoundTrip_ShouldSucceed()
        {
            byte[] encrypted = GcmCryptoHelper.Encrypt(message, key);
            Assert.NotNull(encrypted);
            Assert.True(encrypted.Length > message.Length);

            byte[] decrypted = GcmCryptoHelper.Decrypt(encrypted, key);
            Assert.Equal(message, decrypted);
        }

        [Fact]
        public void Decrypt_ShouldFail_OnTamperedCiphertext()
        {
            byte[] encrypted = GcmCryptoHelper.Encrypt(message, key);
            encrypted[encrypted.Length - 1] ^= 0x01; // flip last byte

            Assert.Throws<AuthenticationTagMismatchException>(() => {
                GcmCryptoHelper.Decrypt(encrypted, key);
            });
        }

        [Fact]
        public void Decrypt_ShouldFail_IfTooShort()
        {
            byte[] invalid = new byte[5];
            Assert.Throws<ArgumentException>(() =>
            {
                GcmCryptoHelper.Decrypt(invalid, key);
            });
        }

        [Fact]
        public void Base64EncodeDecode_ShouldMatchOriginalKey()
        {
            string encoded = GcmCryptoHelper.EncodeBase64Key(key);
            byte[] decoded = GcmCryptoHelper.DecodeBase64Key(encoded);

            Assert.Equal(key, decoded);
        }

        [Fact]
        public void GenerateBase64Key_ShouldReturnValid256BitKey()
        {
            string base64Key = GcmCryptoHelper.GenerateBase64Key();
            byte[] keyBytes = GcmCryptoHelper.DecodeBase64Key(base64Key);

            Assert.Equal(32, keyBytes.Length);
        }

        [Fact]
        public void Encrypt_ShouldGenerateDifferentIVs()
        {
            byte[] key = new byte[32];
            RandomNumberGenerator.Fill(key);

            byte[] c1 = GcmCryptoHelper.Encrypt(message, key);
            byte[] c2 = GcmCryptoHelper.Encrypt(message, key);

            Assert.NotEqual(Convert.ToBase64String(c1), Convert.ToBase64String(c2));
        }

        [Fact]
        public void ReconciliationTest_EncryptWithFixedIv()
        {
            byte[] key = new byte[16]; // match Java
            byte[] iv = new byte[12];  // match Java
            byte[] data = Encoding.UTF8.GetBytes("VertexCacheGCMTest");

            byte[] ciphertext = new byte[data.Length];
            byte[] tag = new byte[16];

            using (AesGcm aes = new AesGcm(key))
            {
                aes.Encrypt(iv, data, ciphertext, tag);
            }

            byte[] combined = new byte[iv.Length + ciphertext.Length + tag.Length];
            Buffer.BlockCopy(iv, 0, combined, 0, iv.Length);
            Buffer.BlockCopy(ciphertext, 0, combined, iv.Length, ciphertext.Length);
            Buffer.BlockCopy(tag, 0, combined, iv.Length + ciphertext.Length, tag.Length);

            // Now decrypt
            byte[] output = GcmCryptoHelper.Decrypt(combined, key);
            Assert.Equal(data, output);

            Console.WriteLine("[RECON] Plaintext: " + Encoding.UTF8.GetString(data));
            Console.WriteLine("[RECON] Key (hex): " + BitConverter.ToString(key).Replace("-", "").ToLower());
            Console.WriteLine("[RECON] IV (hex): " + BitConverter.ToString(iv).Replace("-", "").ToLower());
            Console.WriteLine("[RECON] Encrypted (hex): " + BitConverter.ToString(combined).Replace("-", "").ToLower());
        }
    }
}
